### ISSUE TITLE
(Re-)Implement VAPID WebPush Notification

### ISSUE THREAD

**Issue by genofire** [bromite/bromite#1933]
### Preliminary checklist

- [X] I have read the [README](https://github.com/bromite/bromite/blob/master/README.md)
- [X] I have read the [FAQs](https://github.com/bromite/bromite/blob/master/FAQ.md).
- [X] I have searched [existing issues](https://github.com/bromite/bromite/issues) for my feature request. This is a new issue (NOT a duplicate) and is not related to another issue.

### Is your feature request related to privacy?

Yes

### Is there a patch available for this feature somewhere?

No

### Describe the solution you would like

WebPush Notification wakes up an Browser by a Page's WebService.

This feature is only in Browsers with binary blobs in it, to handle the Connection to Browser Developers Push Server. It would be nice, that the [UnifiedPush](https://unifiedpush.org/) implementation could be used on Android site, so that the User could discide the connection to the Smartphone.

If VAPID could not use the Server implementation, I am willing to develop an vapid to unifiedpush gateway.

### Describe alternatives you have considered

Disassembly the implementation of chrome or Firefox and use there server infrastructure.

**Comment by kokukyocho** [comment]
Hello @genofire!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by csagan5** [comment]
> ### Is your feature request related to privacy?
> Yes

It is not?

> If VAPID could not use the Server implementation, I am willing to develop an vapid to unifiedpush gateway.

What would an unifiedpush gateway look like?

Perhaps you could look into https://microg.org/ instead?

**Comment by genofire** [comment]
Oh i mix up the the wording - i mean an "rewrite proxy" - which reformat the messages from app server for the unifiedpush push provider.

On unified push the server implementation just take an POST to `https://distributor/UP` with an token maybe by query `?token=XX` (the complete URL is generated and response to an APP when the app subscribe for a push notification), and the body is transmitted to the application.

i believe this FAQ has an usefull flow diagram: https://unifiedpush.org/users/faq/#basic-description-of-how-unifiedpush-works and the SPECS are also really short and helpfull.

Yes, of course you could look into microg, but they just try to re-implement FCM (and still use the google server ...).

**Comment by csagan5** [comment]
How does Firefox handle this on Android? An extension?

**Comment by genofire** [comment]
I do not know - just use Fennec from F-Droid, which also has no WebPush support - on Desktop it has they own push server where firefox keeps an connection ...

**Comment by csagan5** [comment]
As I understand this is how WebPush should work: https://web.dev/push-notifications-web-push-protocol/

If we were to have this functionality in Bromite it should not be tied to any specific cloud service but rather allow using any push server, so UnifiedPush looks like covering that functionality via libraries but only in the case "Self-host the server (Difficult)" or anyways if the choice is left to user about which server to use.

> If VAPID could not use the Server implementation, I am willing to develop an vapid to unifiedpush gateway.

I am not understanding very well what you mean here, can you explain? UnifiedPush does not support VAPID and would need a separate gateway that does not currently exist?

**Comment by easrng** [comment]
UnifiedPush has its own protocol (with no encryption) but it looks like it *might* be compatible with VAPID since it basically just delivers POST request bodies to the app. Bromite would need to decrypt the messages though.

**Comment by easrng** [comment]
Ok never mind. VAPID requires headers so there would need to be a shim server that serializes the needed headers and the body into one UnifiedPush message. Keep in mind that VAPID messages are encrypted, so neither the proxy nor the UnifiedPush server would be able to read the notifications.

**Comment by genofire** [comment]
Correct, the messages should be transported encrypted from application server to bromite over the UnifiedPush and Proxy infrastructure.

Yes there should be a litte proxy, which capsulate the headers (and the request body) to an body

**Comment by csagan5** [comment]
If someone would like to do a reference implementation of the server-side application and contribute the patches for Bromite (without any specific default server in use), I'd be happy to consider this for inclusion.

Until then I am afraid this is not planned; I will add the appropriate label and close for now.

Thanks everyone!

**Comment by genofire** [comment]
Could you assign me this issue? Or create one New for developing an reference Server?

So i will Start to develop in two month, but current i am busy. (If it is okay for this project to have that issue for long time).
